### PR TITLE
allow formatErrorMessage callback

### PR DESCRIPTION
**Pull request by chirag04**
something like this can be useful for logging the cost. wdyt?

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -149,7 +149,9 @@ export function createComplexityLimitRule(maxCost, options = {}) {
         }
 
         if (node.kind === 'Document') {
-          if (visitor.getCost() > maxCost) {
+          const cost = visitor.getCost();
+          if (options.onCost) options.onCost(cost);
```
Comment (chirag04) [inline src/index.js]: maybe also pass the query as arg to onCost?

**Comment by taion** [comment]
I'd rather have some callback for formatting the error message to send up to the client.

**Comment by chirag04** [comment]
ok. can you elaborate on the api? `getErrorMessage(cost, documentNode)` maybe?

**Comment by taion** [comment]
just `formatErrorMessage(cost)` perhaps

**Comment by taion** [comment]
Wait, is the idea here that you want to log all costs on queries? Or what's the goal here? Maybe it'd be better to re-use the visitor to create a different rule that only does logging?

**Comment by chirag04** [comment]
initially i thought about logging all costs but figured it makes sense to do it only when validation fails. no? so logging when validation fails is the goal.

**Comment by taion** [comment]
I think there are two separate concerns – what to do right now in dev to figure out how to set cost limits, then what we'll want to do in a production deploy.

**Comment by taion** [comment]
Thinking about this some more, probably https://github.com/4Catalyzer/graphql-validation-complexity/pull/3/commits/4226b5840824d63f46e724494765437256e5c00a is the right way to go.

**Comment by chirag04** [comment]
`it'd be better to re-use the visitor to create a different rule that only does logging`. idk how that would work.

@taion maybe do both. onCost and formatErrorMessage. onCost solves all our usecases but i can imagine somebody needing formatErrorMessage. wdyt?

**Comment by taion** [comment]
Sure – still could be useful to tell without e.g. looking at server logs more details on queries that exceed the cost limit. You'd only want to turn it on in dev, though – would probably be a security issue to enable it in prod.

File: src/index.js (hunk 1 above)
Comment (chirag04) [inline src/index.js]: I still think passing the document node might be useful. we maybe want to log queries that fail in prod as well?

**Comment by chirag04** [comment]
@taion anything left here?

File: src/index.js [hunk 2]
```diff
@@ -149,9 +149,14 @@ export function createComplexityLimitRule(maxCost, options = {}) {
...
+          if (cost > maxCost) {
+            const errorMessage = options.formatErrorMessage ?
```
Comment (taion) [inline src/index.js]: just default `formatErrorMessage` to soemthing

File: src/index.js [hunk 3]
```diff
@@ -149,9 +149,14 @@ export function createComplexityLimitRule(maxCost, options = {}) {
         }
 
         if (node.kind === 'Document') {
-          if (visitor.getCost() > maxCost) {
+          const cost = visitor.getCost();
+          if (options.onCost) options.onCost(cost);
```
Comment (taion) [inline src/index.js]: split out these properties before passing down to visitor

File: src/index.js [hunk 4]
```diff
@@ -149,9 +149,14 @@ export function createComplexityLimitRule(maxCost, options = {}) {
...
+          if (cost > maxCost) {
+            const errorMessage = options.formatErrorMessage ?
+              options.formatErrorMessage(cost) :
+              `query exceeds complexity limit. Cost: ${cost}`;
```
Comment (taion) [inline src/index.js]: we should absolutely _not_ always return the cost here; the calculation may be a private detail

File: README.md [hunk 5]
```diff
@@ -20,6 +20,8 @@ const ComplexityLimitRule = createComplexityLimitRule(1000, {
   scalarCost: 1,
   objectCost: 10, // Default is 0.
   listFactor: 20, // Default is 10.
+  onCost: (cost) => console.log('total'),
+  formatErrorMessage: (cost) => 'Bad Query',
```
Comment (taion) [inline README.md]: these examples are sort of weak (plus don't follow eslint guidelines). also the graf above needs to explain these options.

**Comment by chirag04** [comment]
@taion updated.

**Pull request merged**